### PR TITLE
Add new field boundary algorithm

### PR DESCRIPTION
**Pull request by Flova**
This adds a field boundary algorithm based on OpenCV resizing and blurring. 
It minimizes the python operations while accepting some OpenCV operation overhead. The approach is about 5 times faster on my PC vs. the normal reversed field boundary algorithm. The performance on the Jetson has to be evaluated tho.

### CODE REVIEW

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py [hunk 1]
```diff
@@ -629,13 +657,59 @@ def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps,
...
+                if subsampled_mask[max_y, x_position] < int(green_threshold / 1000 * 255):
+                    # Reset to last step
+                    max_y += 1
+                    break
+
+            # Scale the stuff for the original image + black magic
+            field_boundary_points.append(
+                (int((x_position + 0.5) * (field_mask.shape[1] / x_steps)),
+                 int(max_y * (field_mask.shape[0] / y_steps))))
+        # Fix tuff at the image edges
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: "tuff".

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py [hunk 2]
```diff
@@ -629,13 +657,59 @@ def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps,
...
+        # Define the blur kernel
+        kernel = (2 * roi_height + 1, 2 * roi_width + 1)
+        # Blur the downscaled image to fill holes in the field mask
+        subsampled_mask = cv2.GaussianBlur(subsampled_mask, kernel, 0)
+
+        field_boundary_points = []
+
+        # Iterate horizontal over the image
+        for x_position in range(subsampled_mask.shape[1]):
+            # Iterate vertical over the downscaled mask
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: ...horizontally... ...vertically...

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py [hunk 3]
```diff
@@ -629,13 +657,59 @@ def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps,
...
+                    max_y += 1
+                    break
+
+            # Scale the stuff for the original image + black magic
+            field_boundary_points.append(
+                (int((x_position + 0.5) * (field_mask.shape[1] / x_steps)),
+                 int(max_y * (field_mask.shape[0] / y_steps))))
+        # Fix tuff at the image edges
+        field_boundary_points[0] = (0, field_boundary_points[0][1])
+        field_boundary_points[-1] = (field_mask.shape[1]-1, field_boundary_points[0][1])
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: Are you sure that you want `field_boundary_points[0][1] ` here instead of `field_boundary_points[-1][1]` ?

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py [hunk 4]
```diff
@@ -629,13 +657,59 @@ def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps,
...
+        :returns: list of field boundary points
+        """
+        # calculate the field_mask which contains 0 for non-green pixels and 255 for green pixels in the image
+        # index counting up from top to bottom and left to right
+        field_mask = field_color_detector.get_mask_image()
+
+        # Scale the image down
+        subsampled_mask = cv2.resize(field_mask,(x_steps,y_steps), interpolation=cv2.INTER_AREA)
+        # Define the blur kernel
+        kernel = (2 * roi_height + 1, 2 * roi_width + 1)
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: Why is this so large?

File: bitbots_vision/cfg/Vision.cfg [hunk 5]
```diff
@@ -27,6 +27,7 @@ field_boundary_detector_enum = gen.enum(
     gen.const("iteration",str_t, "iteration", "searches from the top until green"),
     gen.const("dynamic",str_t, "dynamic", "uses head_joint_state messages to switch between methods"),
     gen.const("reversed",str_t, "reversed", "searches from the bottom until white"),
+    gen.const("cvreversed",str_t, "cvreversed", "searches from the bottom until white unsing opencv"),
```
[the remark this answers is not in the document]
Comment (Flova) [reply bitbots_vision/cfg/Vision.cfg]: A bit long, but nice idea. Let's go with that.

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py (hunk 4 above)
Comment (Flova) [reply bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: The kernel? This why should it be large? The roi_height is normally a value of 1. The rest of the calculation is added, to get an odd result every time, which is needed for the blur. Also this calculation output is 1, which means no bluring for an roi size of 0 and 1 on each side for roi 1 and so on.

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py (hunk 3 above)
Comment (Flova) [reply bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: Oh yeah I fix it

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py (hunk 4 above)
Comment (NFiedler) [reply bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: First, "one on each side for roi of 1" would mean that a "height" or "width" of 1 leads to an actual height or width of 3, which is larger than 1. But I think that this is the case everywhere. Second, in this case the width of the kernel considers the subsampled pixels which significantly enlarges the corresponding area in the original image.

Comment (Flova) [reply bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: Of cause the area is quite large after the downscaling. But the kernel normally only considers the direct neighbors by setting the value to 1. This works quite well on our settings and is much faster than bluring the original image. Higher values are nessesary if we increase the resolution further and need to blur bigger areas. Of cause height is a bit misleading if we a talk about some radius like thing, but an other parameter would break the interface quite a bit and this one was the closest one. 
If we want to use the real height/width, we could add an integer division on the height/width value first. That would also output only odd values for the kernel size, but parameter values like 2 and 3 would be equal.

File: bitbots_vision/cfg/Vision.cfg (hunk 5 above)
[the remark this answers is not in the document]
Comment (Flova) [reply bitbots_vision/cfg/Vision.cfg]: Done

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py [hunk 6]
```diff
@@ -629,13 +657,59 @@ def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps,
         return _field_boundary_points
 
 
+class DownsamplingReversedFieldBoundaryAlgorithm(FieldBoundaryAlgorithm):
+    @staticmethod
+    def _calculate_field_boundary(image, field_color_detector, x_steps, y_steps, roi_height, roi_width, roi_increase, green_threshold):
+        # type: () -> list
```
Comment (JanGut) [inline bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: This is defiantly wrong.

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py [hunk 7]
```diff
@@ -629,13 +657,59 @@ def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps,
...
+        """
+        Finds the points of the field boundary visible in the image. Uses the reversed method iterating from bottom to
+        top until it finds enough non green points. Useful for when two fields are adjacent to each other.
+
+        :returns: list of field boundary points
+        """
+        # calculate the field_mask which contains 0 for non-green pixels and 255 for green pixels in the image
+        # index counting up from top to bottom and left to right
```
Comment (JanGut) [inline bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: Why is this here?

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py [hunk 8]
```diff
@@ -629,13 +657,59 @@ def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps,
...
+            for y_position in range(subsampled_mask.shape[0]):
+                # Invert the current vertical value
+                max_y = (subsampled_mask.shape[0] - 1) - y_position
+                # Check if we found a pixel under the threshold
+                if subsampled_mask[max_y, x_position] < int(green_threshold / 1000 * 255):
+                    # Reset to last step
+                    max_y += 1
+                    break
+
+            # Scale the stuff
```
[the remark this answers is not in the document]
Comment (Flova) [reply bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: I totally agree with you

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py (hunk 4 above)
Comment (Flova) [reply bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: I added the integer division thing.

File: bitbots_vision/cfg/Vision.cfg [hunk 9]
```diff
@@ -27,6 +27,7 @@ field_boundary_detector_enum = gen.enum(
     gen.const("iteration",str_t, "iteration", "searches from the top until green"),
     gen.const("dynamic",str_t, "dynamic", "uses head_joint_state messages to switch between methods"),
     gen.const("reversed",str_t, "reversed", "searches from the bottom until white"),
+    gen.const("downsampling_reversed",str_t, "downsampling_reversed", "searches from the bottom until white unsing opencv"),
```
Comment (NFiedler) [inline bitbots_vision/cfg/Vision.cfg]: I don't like 'using opencv'

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (NFiedler) [reply bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: Are you serious???

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py [hunk 10]
```diff
@@ -629,13 +657,59 @@ def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps,
...
+            field_boundary_points.append(
+                (int((x_position + 0.5) * (field_mask.shape[1] / x_steps)),
+                 int(max_y * (field_mask.shape[0] / y_steps))))
+        # Fix stuff at the image edges
```
[the remark this answers is not in the document]
Comment (NFiedler) [reply bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: It still does.

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py [hunk 11]
```diff
@@ -557,16 +583,13 @@ def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps,
 class ReversedFieldBoundaryAlgorithm(FieldBoundaryAlgorithm):
     @staticmethod
     def _calculate_field_boundary(_image, _field_color_detector, _x_steps, _y_steps, _roi_height, _roi_width, _roi_increase, _green_threshold):
-        # type: () -> list
```
Comment (NFiedler) [inline bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: instead of removing these, you/someone should correct them. In a separate PR.

File: bitbots_vision/cfg/Vision.cfg (hunk 9 above)
Comment (Flova) [reply bitbots_vision/cfg/Vision.cfg]: Should be better now

File: bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py (hunk 11 above)
Comment (Flova) [reply bitbots_vision/src/bitbots_vision/vision_modules/field_boundary.py]: Yea, but now they are definitely incorrect and only on very few functions in the whole vision. We should add them for each function.
